twisted/test/test_task.py

```diff
@@ -1199,3 +1199,197 @@ def main(reactor):
         exitError = self.assertRaises(
             SystemExit, task.react, main, [], _reactor=r)
         self.assertEqual(0, exitError.code)
+
+
+
+class _DummyException(Exception):
+    """
+    Exception type to be used in testing.
+    """
+
+
+
+class TimeoutDeferredTests(unittest.SynchronousTestCase):
+    """
+    Tests for the function L{task.timeoutDeferred}
+    """
+    def test_propagatesSuccessResultBeforeTimeoutDefaultTranslation(self):
+        """
+        The L{defer.Deferred} callbacks with the result if it succeeds before
+        the timeout (e.g. timing out the L{defer.Deferred} does not obscure
+        the callback value) when using the default translation function.
+        """
+        clock = Clock()
+        d = defer.Deferred()
+        task.timeoutDeferred(d, 10, clock)
+        d.callback("Result")
+        self.assertEqual(self.successResultOf(d), "Result")
+
+        # The timeout never happens - no errback occurs
+        clock.advance(15)
+        self.assertIsNone(self.successResultOf(d))
+
+    def test_propagatesSuccessResultBeforeTimeoutCustomTranslation(self):
+        """
+        The L{defer.Deferred} callbacks with the result if it succeeds before
+        the timeout (e.g. timing out the L{defer.Deferred} does not obscure
+        the callback value) when using a custom translation function.
+        """
+        clock = Clock()
+        d = defer.Deferred()
+        task.timeoutDeferred(
+            d, 10, clock, translateCancellation=lambda v, t: "OVERRIDDEN")
+        d.callback("Result")
+        self.assertEqual(self.successResultOf(d), "Result")
+
+        # The timeout never happens - the overridden result never happens
+        clock.advance(15)
+        self.assertIsNone(self.successResultOf(d))
+
+    def test_propagatesFailureBeforeTimeoutDefaultTranslation(self):
+        """
+        The L{defer.Deferred} errbacks with the failure if it fails before the
+        timeout (e.g. timing out the L{defer.Deferred} does not obscure the
+        errback failure) when using the default translation function.
+        """
+        clock = Clock()
+        d = defer.Deferred()
+        task.timeoutDeferred(d, 10, clock)
+        d.errback(_DummyException("fail"))
+        self.failureResultOf(d, _DummyException)
```

Comment:
I would suggest to have 

error = _DummyException("fail")

failure = self.failureResultOf(d, _DummyException)

self.assertIs(error, failure.value)
```

this should check that the exact exception is forwarded